Kernel/VirtIO/VirtIOQueue.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2020, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <Kernel/SpinLock.h>
+#include <Kernel/VM/MemoryManager.h>
+
+namespace Kernel {
+
+enum class BufferType {
+    DeviceReadable = 0,
+    DeviceWritable = 1
+};
+
+class VirtIOQueue {
+public:
+    VirtIOQueue(u16 queue_size, u16 notify_offset);
+    ~VirtIOQueue();
+
+    bool is_null() const { return !m_region; }
+    u16 notify_offset() const { return m_notify_offset; }
+
+    void enable_interrupts();
+    void disable_interrupts();
+
+    PhysicalAddress descriptor_area() const { return to_physical(m_descriptors); }
+    PhysicalAddress driver_area() const { return to_physical(m_driver); }
+    PhysicalAddress device_area() const { return to_physical(m_device); }
+
+    bool supply_buffer(const u8* buffer, u32 len, BufferType);
+    bool new_data_available() const;
+
+    bool handle_interrupt();
```

Comment:
if this class handles interrupts, maybe it should inherit from `IRQHandler` class?

Reply:
it handles PCI Interrupts, so it handles from PCI::Device (which inherits from IRQHandler), should it inherit it directly as well?

Reply:
I prefer that the model of `PCI::DeviceController` as a the controller of all VirtIOQueue(s), and a VirtIOQueue class that inherits directly from `IRQHandler`. This way, it will be much easier to enable MSIs or MSI-X later.
`PCI::Device` **should be used only** if you know that the device will never use any other source of interrupts except IRQs and also, the device will only handle one IRQ line, which is not the case with VirtIO.

Reply:
I was under the impression that VirtIO devices only use one IRQ line, in what scenarios do they use more than one?

Reply:
Well, the VirtIO specification allows devices to utilize MSI & MSI-X to reduce IRQ sharing. We will eventually want to support MSI (and MSI-X), because it makes things much easier and also much faster than IRQs, because MSIs bypass the IRQ controller and interrupt the targeted CPU directly.

Reply:
I see, the virtio device itself has to receive IRQs itself as well (as one of the irqs is a device config change), so should the the virt device class be an irqhandler as well besides the devicecontroller?